docs/about/The-Engine-architecture.md

```diff
@@ -1,149 +1,324 @@
-
-There is no way for user code to access this thread either via Dart or native plugins. Even the embedder is free to schedule tasks on this thread that are fairly expensive. This won’t cause jank in Flutter applications but may delay having the futures images and other resources be resolved in a timely manner. Even so, it is recommended that custom embedders set up a dedicated thread for this task runner.
+All the task runners mentioned so far have pretty strong restrictions on the
+kinds of operations that can be performed on this. Blocking the platform task
+runner for an inordinate amount of time may trigger the platform's watchdog, and
+blocking either the UI or raster task runners will cause jank in Flutter
+applications. However, there are tasks necessary for the raster thread that
+require doing some very expensive work. This expensive work is performed on the
+IO task runner.
+
+The main function of the IO task runner is reading compressed images from an
+asset store and making sure these images are ready for rendering on the raster
+task runner. To make sure a texture is ready for rendering, it first has to be
+read as a blob of compressed data (typically PNG, JPEG, etc.) from an asset
+store, decompressed into a GPU friendly format and uploaded to the GPU. These
+operations are expensive and will cause jank if performed on the raster task
+runner. Since only the raster task runner can access the GPU, the IO task runner
+components set up a special context that is in the same sharegroup as the main
+raster task runner context. This happens very early during engine setup and is
+also the reason there is a single task runner for IO tasks. In reality, the
+reading of the compressed bytes and decompression can happen on a thread pool.
+The IO task runner is special because access to the context is only safe from a
+specific thread. The only way to get a resource like
+[`ui.Image`](https://docs.flutter.io/flutter/dart-ui/instantiateImageCodec.html)
+is via an async call; this allows the framework to talk to the IO runner so that
+it can asynchronously perform all the texture operations mentioned. The image
+can then be immediately used in a frame without the raster thread having to do
+expensive work.
+
+There is no way for user code to access this thread either via Dart or native
+plugins. Even the embedder is free to schedule tasks on this thread that are
+fairly expensive. This won’t cause jank in Flutter applications but may delay
+having the futures images and other resources be resolved in a timely manner.
+Even so, it is recommended that custom embedders set up a dedicated thread for
+this task runner.
+
+### Flutter Engine Groups
```

Comment:
Perhaps this can be moved up into "Anatomy of Flutter app" or "Architecture overview".
Basically where "the Engine" is introduced it could be further explained that flutter app could have multiple engines(is that right? not sure what "flutter app" refers to in case of "add to app" for example), and they can be created either directly or via engine groups.
Also, it would be helpful to understand whether flutter engine group concept are limited to android and ios or  are available in other embedders too(desktop linux/mac/win, custom) too.

Reply:
I've worded all this information in terms of threading so I think it fits under the "Threading" section.  I think the section above is talking about a typical single engine case which is what the majority of people will be concerned about.  FlutterEngineGroups are inconsequential in that case.

> Also, it would be helpful to understand whether flutter engine group concept are limited to android and ios or are available in other embedders too(desktop linux/mac/win, custom) too.

We do have links to the iOS and Android APIs.  Desktop platforms are experimental so we don't want to mention those.

Reply:
> FlutterEngineGroups are inconsequential in that case.

Engine group _sounds_ like an entity that owns and manages its engines, so it sounds like something that should be introduced at the same time the engine is introduced.
If it's not, if it's purpose is to just provide a context for engine instantiation, it probably worth clarifying.

Threading is one aspect of flutter engine group behavior, dart vm isolate group reuse is another important one(in terms of the kind of messages can be send between isolates in the engine group, gc and concurrency constraints for worker isolates in the group, requirement for all engines in the group to run from the same dart source).

Reply:
> If it's not, if it's purpose is to just provide a context for engine instantiation, it probably worth clarifying.

Done, clarified at the top we are talking about one engine.

> Threading is one aspect of flutter engine group behavior, dart vm isolate group reuse is another important one(in terms of the kind of messages can be send between isolates in the engine group, gc and concurrency constraints for worker isolates in the group, requirement for all engines in the group to run from the same dart source).

I added the note about sharing source code.

I don't think this is the appropriate doc to go into details about dart isolate groups.  It's a dart concept that should be defined in dart documentation and we can link to it.  I have a link already in the text, let me know if there is a better one.